refactor: Use uint16_t instead of unsigned short

renepickhardt opened the pull request:
I wanted to see if the `up for grabs` label works and looked at PR #17822 originally opend by ahook I saw it had many acks for example by jonatack and practicalswift but needed rebasing. 

So I checked out the remote branch rebased it resolved three conflicts and continued the rebase. 

Hope everything is as expected (:

Commit:
refactor: Use uint16_t instead of unsigned short

removed trailing whitespace to make linter happy

practicalswift commented:
Concept ACK: explicit is better than implicit.

FWIW, we are already assuming that `sizeof(uint16_t) == sizeof(unsigned short)` implicitly in multiple places like here …

https://github.com/bitcoin/bitcoin/blob/35ed88f187c9bc7eedc3a1cf12193e0fbf222057/src/serialize.h#L265-L278

… and also explicitly in `assumptions.h` …

https://github.com/bitcoin/bitcoin/blob/35ed88f187c9bc7eedc3a1cf12193e0fbf222057/src/compat/assumptions.h#L51

renepickhardt commented:
btw running `git grep "unsigned short"` results in: 
```
src/secp256k1/src/tests.c:            unsigned short sh;
```
I could make another commit to this branch where I also changed that but I am not sure if the bitcoin repo is the place to patch libsecp256k1 or if that should be done in https://github.com/bitcoin-core/secp256k1

at least the remote file https://github.com/bitcoin-core/secp256k1/blob/master/src/tests.c has 19 contributers according to git blame where in this repo it only has contributes from @sipa and @laanwj

renepickhardt commented on src/qt/optionsmodel.h:
```diff
@@ -6,6 +6,7 @@
 #define BITCOIN_QT_OPTIONSMODEL_H
 
 #include <amount.h>
+#include <cstdint> 
```
fixed

practicalswift commented:
@renepickhardt Would you mind squashing? I think this one should be ready for final review after squash :)

renepickhardt commented:
Ok so I have squashed the two commits and rebased to current master to make merging easier (: 

if there is anything more to do let me know @practicalswift

practicalswift commented:
ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd -- patch looks correct :)

sipsorcery commented:
ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd.

Nice to replace a C header with a C++ header as well.

practicalswift commented:
ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd -- patch looks correct :)

If anyone wants to follow up by checking potentially unintended/unwarranted uses of `long` this command might be helpful to find candidates :)

```
$ git grep -E '[^a-zA-Z_."]long[^a-zA-Z_."]' -- "*.cpp" "*.h" ":(exclude)src/univalue/" ":(exclude)src/qt/" ":(exclude)src/test/" ":(exclude)src/crc32c/" ":(exclude)src/crypto/common.h" ":(exclude)src/randomenv.cpp" | grep -E -v ':.*[/*]' | grep -vE "(too.long|ftell|strtoul|strol|strtol|how long|really long|fseek|as long)" 
src/httpserver.cpp:    int workQueueDepth = std::max((long)gArgs.GetArg("-rpcworkqueue", DEFAULT_HTTP_WORKQUEUE), 1L);
src/httpserver.cpp:    int rpcThreads = std::max((long)gArgs.GetArg("-rpcthreads", DEFAULT_HTTP_THREADS), 1L);
src/rest.cpp:            if (headers.size() == (unsigned long)count)
src/streams.h:        long nLongPos = nPos;
src/txmempool.cpp:                if (coin.IsSpent() || (coin.IsCoinBase() && ((signed long)nMemPoolHeight) - coin.nHeight < COINBASE_MATURITY)) {
src/txmempool.h:    unsigned long size() const
```

luke-jr commented:
Note the libsecp256k1 `unsigned short` does in fact need to be an unsigned short...

luke-jr commented on src/addrdb.cpp:
```diff
@@ -36,7 +37,7 @@ template <typename Data>
-    unsigned short randv = 0;
+    uint16_t randv = 0;
     GetRandBytes((unsigned char*)&randv, sizeof(randv));
     std::string tmpfn = strprintf("%s.%04x", prefix, randv);
```
This is okay only because `strprintf` has some type magic.

laanwj replied:
Yes, this isn't a problem, `strprintf` is type safe. There is never a need for size specifiers.

luke-jr commented on src/net.cpp:
```diff
@@ -110,9 +111,9 @@ void CConnman::AddOneShot(const std::string& strDest)
+uint16_t GetListenPort()
 {
-    return (unsigned short)(gArgs.GetArg("-port", Params().GetDefaultPort()));
+    return (uint16_t)(gArgs.GetArg("-port", Params().GetDefaultPort()));
```
Can we just drop the cast entirely? Not sure I see the point...

practicalswift replied:
The point is to make a potentially sign-changing narrowing conversion explicit.

`GetArg` returns `int64_t` so it makes sense to make the implicit conversion to `uint16_t` explicit using a cast.

luke-jr commented on src/netbase.cpp:
```diff
@@ -798,11 +799,11 @@ bool ConnectThroughProxy(const proxyType &proxy, const std::string& strDest, int
         static std::atomic_int counter(0);
         random_auth.username = random_auth.password = strprintf("%i", counter++);
-        if (!Socks5(strDest, (unsigned short)port, &random_auth, hSocket)) {
+        if (!Socks5(strDest, (uint16_t)port, &random_auth, hSocket)) {
```
Why not change the type of `port`?

hebasto reviewed (commented):
Approach ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd.

In addition to @luke-jr's [comment](https://github.com/bitcoin/bitcoin/pull/19314#discussion_r444596515), could the type of the `port` parameter in the `Socks5()` function be changed from `int` to `uint16_t`?

laanwj commented:
ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd
After this, the only mention of `unsigned short` left is in the secp256k1 tests, which should not be changed here

> In addition to @luke-jr's comment, could the type of the port parameter in the Socks5() function be changed from int to uint16_t?

I understand why @renepickhardt  didn't do this. Right now, this is a pure refactor, only changing a type to a typedef of the same type (for, in practice, all architectures supported by bitcoin). Changing the parameter type would be an actual code change.

hebasto reviewed (approved):
> > In addition to @luke-jr's comment, could the type of the port parameter in the Socks5() function be changed from int to uint16_t?
> 
> I understand why @renepickhardt didn't do this. Right now, this is a pure refactor, only changing a type to a typedef of the same type (for, in practice, all architectures supported by bitcoin). Changing the parameter type would be an actual code change.

Agree.

ACK 1cabbddbca615b26aa4510c75f459c28d6fe0afd, I have reviewed the code and it looks OK, I agree it can be merged.
